Refactor the per-cluster NAT maps manager

giorio94 opened the pull request:
Let's simplify the per-cluster NAT maps management logic, removing the dependency on the global variable (which does not play well with the hive framework) and splitting the creation/removal tasks from the retrieval of inner maps.

While being at it, let's also slightly improve the performance around these operations by avoiding to attempt opening the inner maps when not strictly necessary.

Finally, let's also move the fake implementation for testing purposes to a separate package, for better separation.

giorio94 committed:
maps: refactor per-cluster NAT maps manager

Let's simplify the per-cluster NAT maps management logic, removing the
dependency on the global variable (which does not play well with the hive
framework) and splitting the creation/removal tasks from the retrieval
of inner maps.

While being at it, let's also slightly improve the performance around
these operations by avoiding to attempt opening the inner maps when
not strictly necessary.

Finally, let's also move the fake implementation for testing purposes to
a separate package, for better separation.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
maps: use typed constants for IP family in NAT maps

Currently, IPv4 and IPv6 NAT maps are distinguished using untyped
boolean constants. Let's replace that with an ad-hoc IPFamily type,
and the corresponding IPv4 and IPv6 constants, for additional clarity.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

gentoo-root commented on cilium/cmd/bpf_nat_list.go:
```diff
@@ -32,17 +31,12 @@ var bpfNatListCmd = &cobra.Command{
 				cmd.PrintErrf("Invalid ClusterID: %s", err.Error())
 				return
 			}
-			ipv4, ipv6, err := nat.ClusterMaps(uint32(clusterID), true, false)
```
Not sure I understand this change... It looks as a behavioral change, not purely refactoring. Is it really in scope of this commit?

Is it related to this?
>While being at it, let's also slightly improve the performance around
>these operations by avoiding to attempt opening the inner maps when
>not strictly necessary.

But it used to be constant false, not true.

giorio94 replied:
Yes, that's a slight behavioral change. The rationale is to align the `cilium bpf nat list cluster ID` command behavior with that of the one used to retrieve global NAT maps, that is showing also IPv6 maps if that address family is enabled. I've moved this change to a separate commit for additional clarity.

giorio94 committed:
cli: retrieve IPv6 NAT per-cluster maps if IPv6 is enabled

Currently, the `bpf nat list cluster ID` command always return only IPv4
maps. Let's additionally retrieve the IPv6 ones when this address family
is enabled, for consistency with the NAT global maps retrieval.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

derailed commented on pkg/maps/nat/types.go:
```diff
@@ -12,6 +12,20 @@ import (
 	"github.com/cilium/cilium/pkg/u8proto"
 )
 
+type IPFamily bool
```
nit: comment

derailed commented on pkg/maps/nat/types.go:
```diff
@@ -12,6 +12,20 @@ import (
+type IPFamily bool
+
+const (
+	IPv4 = IPFamily(true)
```
nit: comment

derailed commented on pkg/maps/nat/per_cluster_nat.go:
```diff
@@ -15,49 +14,86 @@ import (
+var ClusterOuterMapName = clusterOuterMapName
 
-// An interface to interact with the global map.
+func clusterOuterMapName(family IPFamily) string {
```
Looks a bit cumbersome doing the same thing over and over aka if ipv4 else ...  and xxxV4() xxxV6() parity with for the most part very similar logic?? 
Could be out of scope for this PR but seems leveraging generics would help cleaning this further??

giorio94 replied:
Yeah, I agree that there's room for improvement there, especially to reduce code duplication. I would postpone that refactoring though, as this PR is already pretty large, and most occurrences are in the `nat.go` file, which is only touched tangentially by this PR (to replace the ipv4 boolean with the IPFamily constants).

derailed commented on pkg/maps/nat/fake/per_cluster.go:
```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package nat
```
Guessing this is used for testing? If so might make sense to only lasso this in while in test mode and thus use package nat_test instead.

giorio94 replied:
Yes, this is the fake implementation provided for testing purposes. AFAIU it is not possible to import the methods defined inside a `something_test.go` file from another package, so that would not work. Still, this PR already moves the fake implementation to a separate package (under the `fake` directory), so that it can be imported separately from the one with the real implementation.

derailed reviewed (changes requested):
@giorio94 Nice work Marco! Just a few small picks and comment...

derailed reviewed (approved):
@giorio94 Thanks for the updates Marco! LGTM
